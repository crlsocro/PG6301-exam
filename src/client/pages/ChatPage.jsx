import * as React from "react";
import { useEffect, useState } from "react";
import * as ReactDOM from "react-dom";
import { ChatView } from "../views/ChatView";

export function ChatPage() {
    const [username, setUsername] = useState();
    if (!username) {
        return <ChatLoginPage onLogin={(username) => setUsername(username)} />;
    }

    return <ChatView username={username} />;
}

function ChatLoginPage({ onLogin }) {
    const [username, setUsername] = useState("");
    function handleSubmit(e) {
        e.preventDefault();
        onLogin(username);
    }
    return (
        <div>
            <h1>Please log in</h1>
            <form onSubmit={handleSubmit}>
                <input
                    type="text"
                    value={username}
                    onChange={(e) => setUsername(e.target.value)}
                />
                <button>Login</button>
            </form>
        </div>
    );
}