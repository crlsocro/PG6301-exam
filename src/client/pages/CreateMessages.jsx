import React, {useState} from "react";

import { Link } from "react-router-dom";
import {useLoading} from "../views/useLoading";
import {ErrorView} from "../views/ErrorView";
import {LoadingView} from "../views/LoadingView";
import {InputField} from "../views/InputField";
import {ChatView} from "../views/Chatview";

export function CreateMessages({userApi, messageApi}) {
    const [text, setText] = useState("");
    const { data: users, error, loading, reload } = useLoading(
        async () => await userApi.listUsers()
    );

    if (error) {
        return <ErrorView error={error} reload={reload} />;
    }

    if (loading || !users) {
        return <LoadingView />;
    }

    async function submit(e) {
        e.preventDefault();
        await messageApi.createMessage({text});
    }
    return (
        <>
            <h1>Send Message</h1>
            <select
                onChange={(e) => (e.target.value)}
            >
            {users.map(({ id, firstName, lastName}) => (
                    <option value={`/users/${id}`}>{firstName} {lastName}</option>
            ))}
            </select>
            <form onSubmit={submit}>
            <InputField label={"Message"} value={text} onChangeValue={setText} />
            <button>Submit</button>
            </form>
        </>
    );

}
