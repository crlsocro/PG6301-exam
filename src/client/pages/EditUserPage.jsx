import React, { useState } from "react";
import { useParams } from "react-router";
import {useLoading} from "../views/useLoading";
import {ErrorView} from "../views/ErrorView";
import {LoadingView} from "../views/LoadingView";
import {InputField} from "../views/InputField";

function EditUserForm({user, onSubmit}){
    const [firstName, setFirstName] = useState(user.firstName);
    const [lastName, setLastName] = useState(user.lastName);
    const [email, setEmail] = useState(user.email);

    async function submit(e){
        onSubmit(e, {firstName, lastName, email});
    }
    return (
        <form onSubmit={submit}>
            <InputField label={"First Name"} value={firstName} onChangeValue={setFirstName} />
            <InputField label={"Last Name"} value={lastName} onChangeValue={setLastName} />
            <InputField label={"Email"} value={email} onChangeValue={setEmail} />
            <button>Submit</button>
            <h4>Edit an existing user ({firstName} {lastName},{email})</h4>
        </form>
    );
}


export function EditUserPage({userApi}) {
    const {id} = useParams();
    const { data: user, loading, error, reload } = useLoading(
        async () => await userApi.getUser(id),
        [id]
    );
    async function handleSubmit(e, { firstName, lastName, email }) {
        e.preventDefault();
        await userApi.updateUser(id, {firstName, lastName, email});
    }

    if (error) {
        return <ErrorView error={error} reload={reload} />;
    }

    if (loading || !user) {
        return <LoadingView />;
    }

    return <EditUserForm user={user} onSubmit={handleSubmit}/>;
}
