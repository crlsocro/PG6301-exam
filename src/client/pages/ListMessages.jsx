import React from "react";
import { Link } from "react-router-dom";
import {useLoading} from "../views/useLoading";
import {ErrorView} from "../views/ErrorView";
import {LoadingView} from "../views/LoadingView";

export function ListMessages({messageApi}) {
    const { data: messages, error, loading, reload } = useLoading(
        async () => await messageApi.listMessage()
    );

    if (error) {
        return <ErrorView error={error} reload={reload} />;
    }

    if (loading || !messages) {
        return <LoadingView />;
    }

    return (
        <>
            <h1>List Messages</h1>
            {messages.map(({ id, text}) => (
                <li key={id}>
                    <Link to={`/messages/${id}`}>{text}</Link>
                </li>
            ))}
        </>
    );
}