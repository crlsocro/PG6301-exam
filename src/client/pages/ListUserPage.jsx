import React from "react";
import { Link } from "react-router-dom";
import {useLoading} from "../views/useLoading";
import {ErrorView} from "../views/ErrorView";
import {LoadingView} from "../views/LoadingView";



export function ListUserPage({userApi}) {
    const { data: users, error, loading, reload } = useLoading(
        async () => await userApi.listUsers()
    );

    if (error) {
        return <ErrorView error={error} reload={reload} />;
    }

    if (loading || !users) {
        return <LoadingView />;
    }

    return (
        <>
            <h1>List Users</h1>
            {users.map(({ id, firstName, lastName, email }) => (
                <li key={id}>
                    <Link to={`/users/${id}/edit`}>{firstName} {lastName}, {email}</Link>
                </li>
            ))}
        </>
    );
}