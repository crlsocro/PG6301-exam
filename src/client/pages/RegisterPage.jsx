import React, { useState } from "react";
import {InputField} from "../views/InputField";

export function RegisterPage({userApi}) {
    const [firstName, setFirstName] = useState("");
    const [lastName, setLastName] = useState("");
    const [email, setEmail] = useState("");

    async function submit(e) {
        e.preventDefault();
        await userApi.createUser({firstName, lastName, email});
    }
    return (
        <form onSubmit={submit}>
            <h4>Register a user: ({firstName} {lastName},{email})</h4>
            <br/>
            <InputField label={"First Name"} value={firstName} onChangeValue={setFirstName} />
            <InputField label={"Last Name"} value={lastName} onChangeValue={setLastName} />
            <InputField label={"Email"} value={email} onChangeValue={setEmail} />
            <button>Submit</button>
        </form>
    );
}