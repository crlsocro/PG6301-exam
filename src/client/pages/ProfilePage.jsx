import * as React from "react";
import { useLoading } from "../views/useLoading";
import {BrowserRouter, Link} from "react-router-dom";
import {Route, Switch} from "react-router";
import {ErrorView} from "../views/ErrorView";
import {LoadingView} from "../views/LoadingView";
import {ListUserPage} from "./ListUserPage";

export function ProfilePage({ loadProfile }) {
    const {loading, error, data} = useLoading(async () => await loadProfile());

    if (loading) {
        return <div>Loading...</div>;
    }
    if (error) {
        return (
            <div>
                <h1>An error occurred</h1>
                <div>{error.toString()}</div>
            </div>
        );
    }
//TODO: Legge til legge til bruker her
        return (
            <div>
                <h1>Profile</h1>
                <div>{data.name}</div>
                <div>{data.pid}</div>
            </div>

        );
}
