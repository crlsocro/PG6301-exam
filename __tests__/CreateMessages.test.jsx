import React from "react";
import ReactDOM from "react-dom";
import {act, Simulate} from "react-dom/test-utils";
import {MemoryRouter} from "react-router";
import {CreateMessages} from "../src/client/pages/CreateMessages";

async function renderForTest(child) {
    const container = document.createElement("div");
    await act(async () => {
        await ReactDOM.render(<MemoryRouter>{child}</MemoryRouter>, container);
    });
    return container;
}
const userApi = {
    listUsers: async () => [{ id: 1, firstName: "Harald", lastName: "Svensken" }],
};

describe("create text on page", () => {
    it("create text", async () =>{
        const createMessage = jest.fn();
        const container = await renderForTest(
            <CreateMessages messageApi={{createMessage}}/>
        );
        Simulate.change(container.querySelector("input"),{
            target: {value: "hei"}
        });
        Simulate.submit(container.querySelector("form"));
        expect(createMessage).toBeCalledWith({
            text: "hei",
        });
    });
    it("show users on dom", async () => {
        const container = document.createElement("div");
        document.body.appendChild(container);
        await act(async () => {
            ReactDOM.render(
                <MemoryRouter>
                    <CreateMessages userApi={userApi}/>
                </MemoryRouter>,
                container
            );
        });

        expect(container.innerHTML).toMatchSnapshot();
        expect(container.querySelector("div").textContent).toEqual(
            "Harald Svensken"
        );
    });
});