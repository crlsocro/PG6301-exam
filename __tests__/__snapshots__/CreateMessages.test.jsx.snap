// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create text on page show users on dom 1`] = `"<h1>List Users</h1><li><a href=\\"/users/1/edit\\">Harald Svensken, </a></li>"`;
